File: lib/transit/track_chainage.rb
# frozen_string_literal: true

module Transit
  # Cumulative kilometers along a route polyline for O(log n) position lookup.
  class TrackChainage
    def self.for_route(route)
      new(route).build
    end

    def initialize(route)
      @route = route
    end

    def build
      line = longest_line
      return nil if line.nil? || line.length < 2

      cum = [ 0.0 ]
      (1...line.length).each do |i|
        cum << (cum.last + haversine_km(line[i - 1], line[i]))
      end

      { line: line, cum: cum, length_km: cum.last }
    end

    def self.point_at(chainage, distance_km)
      return nil unless chainage

      line = chainage[:line]
      cum = chainage[:cum]
      return nil if line.blank? || cum.blank?

      dist = distance_km.to_f.clamp(0.0, cum.last)
      return { lat: line[0][1], lng: line[0][0] } if dist <= 0
      return { lat: line[-1][1], lng: line[-1][0] } if dist >= cum.last

      hi = cum.bsearch_index { |value| value >= dist } || (cum.length - 1)
      lo = [ hi - 1, 0 ].max
      span = cum[hi] - cum[lo]
      t = span.positive? ? ((dist - cum[lo]) / span) : 0.0
      a = line[lo]
      b = line[hi]
      {
        lat: a[1] + ((b[1] - a[1]) * t),
        lng: a[0] + ((b[0] - a[0]) * t)
      }
    end

    def self.nearest_distance(chainage, lng, lat)
      return nil unless chainage

      line = chainage[:line]
      cum = chainage[:cum]
      best_d = nil
      best_dist = Float::INFINITY

      (0...(line.length - 1)).each do |i|
        ax, ay = line[i]
        bx, by = line[i + 1]
        dx = bx - ax
        dy = by - ay
        len2 = (dx * dx) + (dy * dy)
        t = len2.positive? ? (((lng - ax) * dx) + ((lat - ay) * dy)) / len2 : 0.0
        t = t.clamp(0.0, 1.0)
        px = ax + (dx * t)
        py = ay + (dy * t)
        dist = Math.sqrt(((lat - py) * 111.0)**2 + ((lng - px) * 101.0)**2)
        next if dist >= best_dist

        best_dist = dist
        seg_km = cum[i + 1] - cum[i]
        best_d = cum[i] + (seg_km * t)
      end

      best_d
    end

    private

    def longest_line
      path = @route.geojson_path.to_s.delete_prefix("/")
      full = Rails.root.join("public", path)
      return nil unless File.file?(full)

      data = JSON.parse(File.read(full))
      lines = []
      Array(data["features"]).each do |feature|
        props = feature["properties"] || {}
        next unless props["feature_type"] == "route" || props["feature_type"] == "express_route"

        geometry = feature["geometry"] || {}
        if geometry["type"] == "LineString"
          lines << geometry["coordinates"] if geometry["coordinates"].is_a?(Array)
        elsif geometry["type"] == "MultiLineString"
          Array(geometry["coordinates"]).each { |coords| lines << coords if coords.is_a?(Array) }
        end
      end
      lines.max_by { |line| line.is_a?(Array) ? line.length : 0 }
    rescue JSON::ParserError, Errno::ENOENT
      nil
    end

    def haversine_km(a, b)
      lat1 = a[1].to_f * Math::PI / 180.0
      lat2 = b[1].to_f * Math::PI / 180.0
      dlat = lat2 - lat1
      dlng = (b[0].to_f - a[0].to_f) * Math::PI / 180.0
      h = Math.sin(dlat / 2)**2 + (Math.cos(lat1) * Math.cos(lat2) * Math.sin(dlng / 2)**2)
      2 * 6371.0 * Math.asin([ Math.sqrt(h), 1.0 ].min)
    end
  end
end

File: app/javascript/transit/track_chainage.js
function haversineKm(a, b) {
  const toRad = Math.PI / 180
  const lat1 = a[1] * toRad
  const lat2 = b[1] * toRad
  const dLat = lat2 - lat1
  const dLng = (b[0] - a[0]) * toRad
  const h = Math.sin(dLat / 2) ** 2 + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) ** 2
  return 2 * 6371 * Math.asin(Math.min(1, Math.sqrt(h)))
}

export function buildChainage(line) {
  if (!Array.isArray(line) || line.length < 2) return null

  const cum = [ 0 ]
  for (let i = 1; i < line.length; i += 1) {
    cum.push(cum[i - 1] + haversineKm(line[i - 1], line[i]))
  }

  return { line, cum, lengthKm: cum[cum.length - 1] }
}

export function longestTrackLine(tracks) {
  if (!Array.isArray(tracks) || tracks.length === 0) return null
  return tracks.reduce((best, line) => {
    if (!Array.isArray(line) || line.length < 2) return best
    if (!best || line.length > best.length) return line
    return best
  }, null)
}

export function pointAtDistance(chainage, distanceKm) {
  if (!chainage?.line || !chainage?.cum?.length) return null

  const { line, cum } = chainage
  const dist = Math.max(0, Math.min(cum[cum.length - 1], Number(distanceKm) || 0))
  if (dist <= 0) return { lat: line[0][1], lng: line[0][0] }
  if (dist >= cum[cum.length - 1]) return { lat: line[line.length - 1][1], lng: line[line.length - 1][0] }

  let lo = 0
  let hi = cum.length - 1
  while (lo < hi) {
    const mid = (lo + hi) >> 1
    if (cum[mid] < dist) lo = mid + 1
    else hi = mid
  }

  const i = Math.max(lo, 1)
  const span = cum[i] - cum[i - 1]
  const t = span > 0 ? (dist - cum[i - 1]) / span : 0
  const a = line[i - 1]
  const b = line[i]
  return {
    lat: a[1] + ((b[1] - a[1]) * t),
    lng: a[0] + ((b[0] - a[0]) * t)
  }
}

export function nearestDistance(chainage, lng, lat) {
  if (!chainage?.line || !chainage?.cum) return null

  const { line, cum } = chainage
  let bestD = null
  let bestDist = Infinity

  for (let i = 0; i < line.length - 1; i += 1) {
    const ax = line[i][0]
    const ay = line[i][1]
    const bx = line[i + 1][0]
    const by = line[i + 1][1]
    const dx = bx - ax
    const dy = by - ay
    const len2 = (dx * dx) + (dy * dy)
    let t = len2 > 0 ? (((lng - ax) * dx) + ((lat - ay) * dy)) / len2 : 0
    t = Math.max(0, Math.min(1, t))
    const px = ax + (dx * t)
    const py = ay + (dy * t)
    const dist = Math.sqrt(((lat - py) * 111) ** 2 + ((lng - px) * 101) ** 2)
    if (dist >= bestDist) continue
    bestDist = dist
    bestD = cum[i] + ((cum[i + 1] - cum[i]) * t)
  }

  return bestD
}
